### PR TITLE
Fix rate estimator bucket race condition

### PR DESCRIPTION
**Pull request by guilload**
### Description
Per title.

### How was this PR tested?
- Added some test cases

### CODE REVIEW

File: quickwit/quickwit-common/src/tower/rate_estimator.rs [hunk 1]
```diff
@@ -119,59 +124,57 @@ impl RateEstimator for SmaRateEstimator {
...
-    fn cumulated_work(&self) -> u64 {
-        self.cumulated_work.load(Ordering::Relaxed)
+    fn work(&self) -> u64 {
+        self.bits.load(Ordering::Relaxed) & Self::WORK_MASK
     }
 
-    fn update(&self, work: u64) {
-        self.cumulated_work.fetch_add(work, Ordering::Relaxed);
-    }
+    fn increment_work(&self, work: u64, bucket_ord: u64, num_buckets: u64) {
```
Comment (guilload) [inline quickwit/quickwit-common/src/tower/rate_estimator.rs:153]: This should be `self.bits.fetch_add(work & Self::WORK_MASK, Ordering::Relaxed)` but that seems overkill.

File: quickwit/quickwit-common/src/tower/rate_estimator.rs [hunk 2]
```diff
@@ -119,59 +124,57 @@ impl RateEstimator for SmaRateEstimator {
...
 }
 
 impl Bucket {
-    fn new(bucket_ord: u64, cumulated_work: u64) -> Self {
-        Self {
-            bucket_ord: AtomicU64::new(bucket_ord),
-            cumulated_work: AtomicU64::new(cumulated_work),
-        }
-    }
+    const COLOR_MASK: u64 = 0x8000000000000000;
```
Comment (fulmicoton) [inline quickwit/quickwit-common/src/tower/rate_estimator.rs:145]:
```suggestion
    const COLOR_MASK: u64 = 1u64 << 63;
```

File: quickwit/quickwit-common/src/tower/rate_estimator.rs [hunk 3]
```diff
@@ -119,59 +124,57 @@ impl RateEstimator for SmaRateEstimator {
...
 
-    fn bucket_ord(&self) -> u64 {
-        self.bucket_ord.load(Ordering::Acquire)
-    }
+    const WORK_MASK: u64 = 0x7FFFFFFFFFFFFFFF;
```
Comment (fulmicoton) [inline quickwit/quickwit-common/src/tower/rate_estimator.rs:147]:
```suggestion
    const WORK_MASK: u64 = u64::MAX - COLOR_MASK;
```

File: quickwit/quickwit-common/src/tower/rate_estimator.rs (hunk 1 above)
Comment (fulmicoton) [inline quickwit/quickwit-common/src/tower/rate_estimator.rs:153]: We could pass color directly in this method maybe?

**Review by fulmicoton — APPROVED** [review]

File: quickwit/quickwit-common/src/tower/rate_estimator.rs [hunk 4]
```diff
@@ -119,59 +124,57 @@ impl RateEstimator for SmaRateEstimator {
     fn update(&mut self, _started_at: Instant, ended_at: Instant, work: u64) {
```
Comment (fulmicoton) [inline quickwit/quickwit-common/src/tower/rate_estimator.rs:124]: could we ditch `_start_at`? I know it is used in the dummy impl, but do we use it? Could it be internalized in the Dummy impl?

Comment (guilload) [reply quickwit/quickwit-common/src/tower/rate_estimator.rs:119]: I should actually use it... The argument should be used to check whether the duration spans multiple buckets, and if so, we should spread the amount of work evenly between buckets. However, in our case, the buckets are large (10s), and the requests are short, so the approximation seems ok. What do you think?

**Pull request merged**